mw/i2c_service/pca9685/service/pca9685.cpp

```diff
@@ -0,0 +1,156 @@
+/**
+ * @file pca9685.cpp
+ * @author Mateusz Krajewski (matikrajek42@gmail.com)
+ * @brief 
+ * @version 0.1
+ * @date 2024-03-24
+ * 
+ * @copyright Copyright (c) 2024
+ * 
+ */
+
+#include "pca9685.hpp"
+#include <utility>
+
+namespace simba {
+namespace i2c {
+
+namespace {
+    constexpr uint8_t PCA9685_ADDRESS = 0x70;
+    constexpr uint8_t MODE1_REG = 0x00;
+    constexpr uint8_t PRESCALE_REG = 0xFE;
+    constexpr uint8_t LED0_ON_L = 0x06;
+    constexpr uint8_t LED0_ON_H = 0x07;
+    constexpr uint8_t LED0_OFF_L = 0x08;
+    constexpr uint8_t LED0_OFF_H = 0x09;
+
+}
+/**
+ * @brief 
+ * 
+ * @param channel (0-15) 
+ * @param pos (0-4096)
+ * @return core::ErrorCode 
+ */
+core::ErrorCode PCA9685::SetServo(uint8_t channel, uint16_t pos) {
+    std::lock_guard<std::mutex> lock(*this->mtx);
+    int i2cFile;
+    const char *i2cDevice = "/dev/i2c-2";
+    if ((i2cFile = open(i2cDevice, O_RDWR)) < 0) {
+      return core::ErrorCode::kInitializeError;
+    }
+
+    if (ioctl(i2cFile, I2C_SLAVE, PCA9685_ADDRESS) < 0) {
+        return core::ErrorCode::kInitializeError;
+    }
+
+    char buf[2] = {MODE1_REG, 0x01};
+    if (write(i2cFile, buf, 2) != 2) {
+        return core::ErrorCode::kInitializeError;
+    }
+
+    int prescale = 121;
+    buf[0] = PRESCALE_REG;
+    buf[1] = prescale;
+    if (write(i2cFile, buf, 2) != 2) {
+        return core::ErrorCode::kInitializeError;
+    }
+
+    int onValue = 0;
+    buf[0] = LED0_ON_L+4*channel;
+    buf[1] = onValue & 0xFF;
+    if (write(i2cFile, buf, 2) != 2) {
+        return core::ErrorCode::kInitializeError;
+    }
+    buf[0] = LED0_ON_H+4*channel;
+    buf[1] = onValue >> 8;
+    if (write(i2cFile, buf, 2) != 2) {
+        return core::ErrorCode::kInitializeError;
+    }
+    buf[0] = LED0_OFF_L+4*channel;
+    buf[1] = pos & 0xFF;
+    if (write(i2cFile, buf, 2) != 2) {
+        return core::ErrorCode::kInitializeError;
+    }
+    buf[0] = LED0_OFF_H+4*channel;
+    buf[1] = pos >> 8;
+    if (write(i2cFile, buf, 2) != 2) {
+        return core::ErrorCode::kInitializeError;
+    }
+    close(i2cFile);
+    return core::ErrorCode::kOk;
```

Comment:
nie da sie tego lepiej ???

Reply:
Przy zmianie 1 pwm trzeba zmienić 4 rejestry + 2 odpowiadające za wczytanie zmian.

Reply:
no ale to beznadziejnie wygląda

Reply:
zmieniłem troche forme więc może teraz będzie lepiej